src/fabric.c

```diff
@@ -62,6 +63,33 @@ static struct fi_prov *prov_head, *prov_tail;
 static volatile int init = 0;
 static pthread_mutex_t ini_lock = PTHREAD_MUTEX_INITIALIZER;
 
+static char **prov_incl = NULL;
+static char **prov_excl = NULL;
+
+/* returns 1 if the provider should be kept, 0 if it should be skipped */
+static int filter_provider(struct fi_provider *provider)
+{
+	int keep, i;
+
+	assert(!(prov_incl && prov_excl));
+
+	if (prov_incl != NULL) {
+		keep = 0;
+		for (i = 0; prov_incl[i] != NULL; ++i)
+			if (strcmp(provider->name, prov_incl[i]) == 0)
+				keep = 1;
+		return keep;
```

Comment:
Very minor nit: is there a reason to (assign to keep, complete the loop, and then return keep) vs. (return 1) from the middle of the loop?  If there's no reason and it's just 6 of one / half-dozen of the other, then I'm not advocating switching to the other style.

(analogous comment about the prov_excl loop, below)